examples/decoding/plot_decoding_ssd_csp_eeg.py

```diff
@@ -0,0 +1,255 @@
+from mne.decoding import SSD
+
+###############################################################################
+# Set parameters and read data
+
+# avoid classification of evoked responses by using epochs that start 1s after
+# cue onset.
+tmin, tmax = -1., 4.
+event_id = dict(left=2, right=3)  # Motor imagery: left vs right hand
+subject = 2
+runs = [4, 8, 12]
+
+raw_fnames = eegbci.load_data(subject, runs)
+raw = concatenate_raws([read_raw_edf(f, preload=True) for f in raw_fnames])
+eegbci.standardize(raw)  # set channel names
+montage = make_standard_montage('standard_1005')
+raw.set_montage(montage)
+
+# strip channel names of "." characters
+raw.rename_channels(lambda x: x.strip('.'))
+# for the sake of comparison, we save the raw object before filteing
+raw_ssd = raw.copy()
+##############################################################################
+# Data filtering.
+
+# we are going to filter data in the alpha band
+freq_signal = [8, 12]
+# apply band-pass filter
+raw.filter(freq_signal[0], freq_signal[1], fir_design='firwin',
+           skip_by_annotation='edge')
+
+events, _ = events_from_annotations(raw, event_id=dict(T1=2, T2=3))
+
+picks = pick_types(raw.info, meg=False, eeg=True, stim=False, eog=False,
+                   exclude='bads')
+
+# Extract epochs between 1 and 2s
+epochs = Epochs(raw, events, event_id, tmin, tmax, proj=True, picks=picks,
+                baseline=None, preload=True)
+epochs_cropped = epochs.copy().crop(tmin=1., tmax=2.)
+labels = epochs.events[:, -1] - 2
+epochs_data = epochs_cropped.get_data()
+###############################################################################
+# Traditional CSP+LDA pipeline
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+# define pipelines with monte-carlo simulations
+cv = ShuffleSplit(10, test_size=0.2, random_state=42)
+# pipeline methods
+lda = LinearDiscriminantAnalysis()
+csp = CSP(n_components=4, reg=None, log=True, norm_trace=False)
+pipe_csp = Pipeline([('CSP', csp), ('LDA', lda)])
+# this is the accuracy we would like to improve by adding SSD into the pipeline
+scores_csp = cross_val_score(pipe_csp, epochs_data, labels, cv=cv,
+                             n_jobs=1).mean()
+###############################################################################
+# SSD enhances SNR, thus what is considered 'noise' should be defined
+# Typically the 2 Hz sourronunding frequencies are taken
+freq_noise = [6, 14]
+# when applying SSD, data should be broader band filtered
+# a broader band of the freq_signal is defined
+freq_ssd = [5, 15]
+# as before, filter data
+raw_ssd.filter(freq_ssd[0], freq_ssd[1], fir_design='firwin',
+               skip_by_annotation='edge')
+###############################################################################
+# SSD + CSP + LDA pipeline
+# ^^^^^^^^^^^^^^^^^^^^^^^^^
+
+# SSD can be applied either BEFORE or AFTER data epoching, each approach has
+# it owns advantages and disadvantages. Here both approaches are going to be
+# implemented.
```

Comment:
I would argue that doing it outside the pipeline can cause a potential overfit, although maybe minimal
as it's unsupervised. I would not advertise this.

Reply:
Exactly!! that is my maaaain concern!!! I am showing this because for other non-machine learning applications may have sense to do it before epoching, but I wouldn't be happy if this can be misleading and give the wrong message regarding how a nested ML pipeline should be trained and evaluated. So, maybe remove it? or add a warning?